File: LightWave2D/grid.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Union, Optional
import numpy
from pydantic.dataclasses import dataclass
import shapely.geometry as geo
from TypedUnit import Length

from LightWave2D.physics import Physics
from LightWave2D.utils import config_dict


class NameSpace:
    """
    A class to dynamically create attributes from keyword arguments.
    """

    def __init__(self, **kwargs):
        for k, v in kwargs.items():
            setattr(self, k, v)


@dataclass(config=config_dict, kw_only=True)
class Grid:
    """
    Represents a 2D simulation grid with specified dimensions and time steps.

    Attributes
    ----------
    resolution : Length
        The spatial resolution of the grid.
    size_x : Length
        The size of the grid in the x-direction.
    size_y : Length
        The size of the grid in the y-direction.
    n_steps : int
        The number of time steps for the simulation (default is 200).
    """
    resolution: Length
    size_x: Length
    size_y: Length
    n_steps: int = 200

    def __post_init__(self):
        self.resolution = self.resolution
        self.size_x = self.size_x
        self.size_y = self.size_y

        self.n_x = int(self.size_x / self.resolution)
        self.n_y = int(self.size_y / self.resolution)
        self.dx = self.size_x / self.n_x  # Should be approximately equal to the resolution
        self.dy = self.size_y / self.n_y  # Should be approximately equal to the resolution
        self.dt = 1 / (Physics.c * numpy.sqrt(1 / self.dx**2 + 1 / self.dy**2))  # Time step size using Courant condition

        self.shape = (self.n_x, self.n_y)
        self.time_stamp = numpy.arange(self.n_steps) * self.dt.to('second')
        self.x_stamp = numpy.linspace(0, self.size_x, self.n_x)
        self.y_stamp = numpy.linspace(0, self.size_y, self.n_y)

        self.x_mesh, self.y_mesh = numpy.meshgrid(self.x_stamp, self.y_stamp)

        self.polygon = geo.Polygon([
            (self.x_stamp[0].to('meter').magnitude, self.y_stamp[0].to('meter').magnitude),
            (self.x_stamp[0].to('meter').magnitude, self.y_stamp[-1].to('meter').magnitude),
            (self.x_stamp[-1].to('meter').magnitude, self.y_stamp[0].to('meter').magnitude),
            (self.x_stamp[-1].to('meter').magnitude, self.y_stamp[-1].to('meter').magnitude)
        ]).convex_hull

    def get_distance_grid(self, x0: float = 0, y0: float = 0) -> numpy.ndarray:
        """
        Compute the distance grid from a given point (x0, y0).

        Args:
            x0 (float): x-coordinate of the reference point (default is 0).
            y0 (float): y-coordinate of the reference point (default is 0).

        Returns:
            numpy.ndarray: A 2D array of distances from the reference point.
        """
        x_mesh, y_mesh = numpy.meshgrid(self.x_stamp, self.y_stamp)
        distance_mesh = numpy.sqrt((x_mesh - x0)**2 + (y_mesh - y0)**2)
        return distance_mesh

    def get_coordinate(self, x: Optional[Union[float, str, Length]] = None, y: Optional[Union[float, str, Length]] = None) -> NameSpace:
        """
        Get the coordinate and index for a given position in the grid.

        Parameters
        ----------
        x : float | str | Length
            The x-coordinate or position string ('left', 'center', 'right').
        y : float | str | Length
            The y-coordinate or position string ('bottom', 'center', 'top').

        Returns
        -------
        NameSpace
            An object containing the coordinates and indices.
        """
        coordinate = NameSpace()

        if isinstance(x, str):
            x = self.parse_x_position(x)
        if isinstance(y, str):
            y = self.parse_y_position(y)

        if x is not None:
            x = numpy.clip(x, self.x_stamp[0], self.x_stamp[-1])
            coordinate.x = x
            coordinate.x_index = int(x / self.dx)

        if y is not None:
            y = numpy.clip(y, self.y_stamp[0], self.y_stamp[-1])
            coordinate.y = y
            coordinate.y_index = int(y / self.dy)

        return coordinate

    def parse_y_position(self, value: Union[str, float, Length]) -> float:
        """
        Convert a position string to a y-coordinate.

        Parameters
        ----------
        value : str | float | Length
            The position string or coordinate value.

        Returns
        -------
        float
            The corresponding y-coordinate.
        """
        if isinstance(value, str):
            value = value.lower()
            if '%' in value:
                percentage = float(value.strip('%')) / 100.0
                delta = self.y_stamp[-1] - self.y_stamp[0]

                return self.y_stamp[0] + percentage * delta

            assert value in ['bottom', 'center', 'top'], f"Invalid position: {value}. Valid inputs are ['bottom', 'center', 'top']."
            match value:
                case 'bottom':
                    return self.y_stamp[0]
                case 'center':
                    return numpy.mean(self.y_stamp)
                case 'top':
                    return self.y_stamp[-1]

        return value

    def parse_x_position(self, value: Union[str, float, Length]) -> float:
        """
        Convert a position string to an x-coordinate.

        Parameters
        ----------
        value : str | float | Length
            The position string or coordinate value.

        Returns
        -------
        float
            The corresponding x-coordinate.
        """
        if isinstance(value, str):
            value = value.lower()
            if '%' in value:
                percentage = float(value.strip('%')) / 100.0
                delta = self.x_stamp[-1] - self.x_stamp[0]

                return self.x_stamp[0] + percentage * delta

            assert value in ['right', 'center', 'left'], f"Invalid position: {value}. Valid inputs are ['right', 'center', 'left']."
            match value:
                case 'right':
                    return self.x_stamp[-1]
                case 'center':
                    return numpy.mean(self.x_stamp)
                case 'left':
                    return self.x_stamp[0]

        return value

File: LightWave2D/cpp/grid/grid.cpp
#include "grid.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace py = pybind11;

Grid::Grid(py::object resolution_, py::object size_x_, py::object size_y_, int64_t n_steps_)
    : resolution(std::move(resolution_)), size_x(std::move(size_x_)), size_y(std::move(size_y_)), n_steps(n_steps_) {
    ureg_ = py::module_::import("TypedUnit").attr("ureg");
    const double resolution_m = meter_value(resolution);
    const double size_x_m = meter_value(size_x);
    const double size_y_m = meter_value(size_y);
    if (resolution_m <= 0.0 || size_x_m <= 0.0 || size_y_m <= 0.0)
        throw py::value_error("resolution and grid sizes must be positive.");
    if (n_steps < 1)
        throw py::value_error("n_steps must be at least 1.");
    n_x = static_cast<int64_t>(size_x_m / resolution_m);
    n_y = static_cast<int64_t>(size_y_m / resolution_m);
    if (n_x < 2 || n_y < 2)
        throw py::value_error("Grid dimensions must each contain at least two cells.");

    dx = size_x / py::int_(n_x);
    dy = size_y / py::int_(n_y);
    const double dx_m = meter_value(dx);
    const double dy_m = meter_value(dy);
    dt = meter_quantity(1.0 / (299792458.0 * std::sqrt(1.0 / (dx_m * dx_m) + 1.0 / (dy_m * dy_m)))).attr("to")("meter") / ureg_.attr("meter") * ureg_.attr("second");
    shape = py::make_tuple(n_x, n_y);

    py::object numpy = py::module_::import("numpy");
    time_stamp = numpy.attr("arange")(n_steps) * dt.attr("to")("second");
    x_stamp = numpy.attr("arange")(n_x) * dx;
    y_stamp = numpy.attr("arange")(n_y) * dy;
    py::tuple meshes = numpy.attr("meshgrid")(x_stamp, y_stamp).cast<py::tuple>();
    x_mesh = meshes[0];
    y_mesh = meshes[1];
    py::object polygon_type = py::module_::import("shapely.geometry").attr("Polygon");
    py::object points = py::make_tuple(
        py::make_tuple(0.0, 0.0), py::make_tuple(0.0, meter_value(y_stamp.attr("__getitem__")(-1))),
        py::make_tuple(meter_value(x_stamp.attr("__getitem__")(-1)), 0.0),
        py::make_tuple(meter_value(x_stamp.attr("__getitem__")(-1)), meter_value(y_stamp.attr("__getitem__")(-1))));
    polygon = polygon_type(points).attr("convex_hull");
}

double Grid::meter_value(const py::object& value) const {
    py::object quantity = value;
    if (py::hasattr(quantity, "to")) quantity = quantity.attr("to")("meter").attr("magnitude");
    return quantity.cast<double>();
}

py::object Grid::meter_quantity(double value) const { return py::float_(value) * ureg_.attr("meter"); }

py::object Grid::parse_position(py::object value, bool is_x) const {
    if (!py::isinstance<py::str>(value)) return value;
    const std::string text = py::str(value).cast<std::string>();
    const py::object stamp = is_x ? x_stamp : y_stamp;
    if (text.find('%') != std::string::npos) {
        const double percentage = std::stod(text.substr(0, text.size() - 1)) / 100.0;
        return stamp.attr("__getitem__")(0) + py::float_(percentage) * (stamp.attr("__getitem__")(-1) - stamp.attr("__getitem__")(0));
    }
    if (is_x) {
        if (text == "left") return stamp.attr("__getitem__")(0);
        if (text == "center") return py::module_::import("numpy").attr("mean")(stamp);
        if (text == "right") return stamp.attr("__getitem__")(-1);
        throw py::value_error("Invalid x position. Use left, center, right, or a percentage.");
    }
    if (text == "bottom") return stamp.attr("__getitem__")(0);
    if (text == "center") return py::module_::import("numpy").attr("mean")(stamp);
    if (text == "top") return stamp.attr("__getitem__")(-1);
    throw py::value_error("Invalid y position. Use bottom, center, top, or a percentage.");
}

py::object Grid::parse_x_position(py::object value) const { return parse_position(std::move(value), true); }
py::object Grid::parse_y_position(py::object value) const { return parse_position(std::move(value), false); }

py::object Grid::get_coordinate(py::object x, py::object y) const {
    py::object namespace_ = py::module_::import("types").attr("SimpleNamespace")();
    if (!x.is_none()) {
        py::object x_value = py::module_::import("numpy").attr("clip")(parse_x_position(x), x_stamp.attr("__getitem__")(0), x_stamp.attr("__getitem__")(-1));
        const double x_m = meter_value(x_value);
        namespace_.attr("x") = x_value;
        namespace_.attr("x_index") = std::min(static_cast<int64_t>(x_m / meter_value(dx)), n_x - 1);
    }
    if (!y.is_none()) {
        py::object y_value = py::module_::import("numpy").attr("clip")(parse_y_position(y), y_stamp.attr("__getitem__")(0), y_stamp.attr("__getitem__")(-1));
        const double y_m = meter_value(y_value);
        namespace_.attr("y") = y_value;
        namespace_.attr("y_index") = std::min(static_cast<int64_t>(y_m / meter_value(dy)), n_y - 1);
    }
    return namespace_;
}

py::object Grid::get_distance_grid(py::object x0, py::object y0) const {
    py::tuple meshes = py::module_::import("numpy").attr("meshgrid")(x_stamp, y_stamp).cast<py::tuple>();
    py::object x_mesh_ = meshes[0];
    py::object y_mesh_ = meshes[1];
    return py::module_::import("numpy").attr("sqrt")((x_mesh_ - x0) * (x_mesh_ - x0) + (y_mesh_ - y0) * (y_mesh_ - y0));
}
